### PR TITLE
network: check SR-IOV PF state

### LINKED ISSUE #23315: systemd-networkd: Cannot create vlan network on SR-IOV VF after recent update
Hello, I am on Arch Linux 5.17.37-1-lts:

```
# systemctl --version
systemd 250 (250.5-1-arch)
```

~~I think this began with the upgrade from 250.4 to 250.5 but I am not sure.~~

The update was from 249.5-1 to 250.5-1:

```
[2022-05-05T23:36:54+0300] [ALPM] upgraded systemd-libs (249.5-1 -> 250.5-1)
[2022-05-05T23:36:55+0300] [ALPM] upgraded systemd (249.5-1 -> 250.5-1)
[2022-05-05T23:38:07+0300] [ALPM] upgraded systemd-sysvcompat (249.5-1 -> 250.5-1)
```

I am unable to create vlan networks on SR-IOV VF interfaces. I am using SR-IOV VFs and their VLAN networks for docker networking on the host. Journalctl shows this error:

```
May 08 23:36:27 hostname systemd-networkd[3798258]: eno1v16: Could not create stacked netdev: Input/output error
May 08 23:36:27 hostname systemd-networkd[3798258]: eno1v16: Failed
```

Manually creating the vlan interface with:

```
ip link add link eno1v16 name eno1v16.400 type vlan id 400
```

works as expected. The interface gets created and is functional.

### PR DESCRIPTION
**Pull request by yuwata**
Hopefully fixes #23315.

### CODE REVIEW

File: src/network/networkd-link.h [hunk 1]
```diff
@@ -52,6 +52,7 @@ typedef struct Link {
         int ifindex;
         int master_ifindex;
         int dsa_master_ifindex;
+        int phys_port_ifindex; /* for SR-IOV */
```
Comment (keszybz) [inline src/network/networkd-link.h]: Maybe add a comment here that this is initially 0, and will later be set to either -1 or positive?

File: src/network/networkd-sriov.c [hunk 2]
```diff
@@ -103,3 +107,124 @@ int link_request_sr_iov_vfs(Link *link) {
...
+                return r;
+
+        dir = opendir(pci_physfn_syspath);
+        if (!dir)
+                return -errno;
+
+        FOREACH_DIRENT_ALL(de, dir, break) {
+                _cleanup_free_ char *virtfn_link_file = NULL, *virtfn_pci_syspath = NULL;
+
+                if (isempty(startswith(de->d_name, "virtfn")))
```
Comment (keszybz) [inline src/network/networkd-sriov.c:268]: Is it intentional to check if the suffix is non-empty? `!startswith(…)` would be enough if not.

**Review by keszybz — COMMENTED** [review]
Looks reasonable. Let's maybe wait for confirmation from @sotiris-bos .

File: src/network/networkd-sriov.c (hunk 2 above)
Comment (yuwata) [reply src/network/networkd-sriov.c:268]: All links are suffixed with a number, e.g. `virtfn0`, hence I think it is safer to ignore the file `virtfn`.

File: src/network/networkd-link.h (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-link.h]: Sure, will do so after it is confirmed that this PR works well.

File: src/network/networkd-sriov.c [hunk 3]
```diff
@@ -103,3 +107,124 @@ int link_request_sr_iov_vfs(Link *link) {
...
+                        continue;
```
Comment (poettering) [inline src/network/networkd-sriov.c:269]: but this whole check can be rewritten as `if (streq(de->d_name, "virtfn"))`, no?

Comment (keszybz) [reply src/network/networkd-sriov.c:269]: No. We want to filter out e.g. "foobar".

**Comment by yuwata** [comment]
From the logs provided in #23315, the first two commits work as expected, but they do not solve the issue completely. I added three more commits, which hopefully fix the issue.

File: src/network/networkd-sriov.c (hunk 3 above)
Comment (poettering) [reply src/network/networkd-sriov.c:269]: i see. 

maybe add comment: "filter out everything that doesn't begin with `virtfn` and is followed by at least one more character". 

(at least I didn't grok what this does at first...)

File: src/network/networkd-sriov.c [hunk 4]
```diff
@@ -103,3 +107,228 @@ int link_request_sr_iov_vfs(Link *link) {
...
+        r = sd_device_get_parent_with_subsystem_devtype(link->sd_device, "pci", NULL, &pci_dev);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_syspath(pci_dev, &pci_syspath);
+        if (r < 0)
+                return r;
+
+        pci_physfn_syspath = strjoina(pci_syspath, "/physfn");
+        r = sd_device_new_from_syspath(&pci_physfn_dev, pci_physfn_syspath);
```
Comment (poettering) [inline src/network/networkd-sriov.c]: we should probably add `sd_device_new_child(sd_device *d, const char *suffix)` or so, which combines the sd_device_get_syspath() + path_join() + sd_device_new_from_syspath() into one. I am pretty sure this is a pretty common pattern in our code...

File: src/network/networkd-sriov.c [hunk 5]
```diff
@@ -103,3 +107,228 @@ int link_request_sr_iov_vfs(Link *link) {
...
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_syspath(pci_physfn_dev, &pci_physfn_syspath);
+        if (r < 0)
+                return r;
+
+        dir = opendir(pci_physfn_syspath);
+        if (!dir)
+                return -errno;
```
Comment (poettering) [inline src/network/networkd-sriov.c:259]: and another helper sd_device_get_children() might make sense too, that returns a char** array of child nodes. i.e. that is a wrapper around readdir_all() for the device, and filters ./.. and everything that isn't a dir or a symlink to a dir.

Comment (poettering) [reply src/network/networkd-sriov.c:259]: (and sd_device_get_children() probably should return normalized sysfs paths?)

File: src/network/networkd-sriov.c [hunk 6]
```diff
@@ -103,3 +107,228 @@ int link_request_sr_iov_vfs(Link *link) {
...
+
+                if (isempty(startswith(de->d_name, "virtfn")))
+                        continue;
+
+                virtfn_link_file = path_join(pci_physfn_syspath, de->d_name);
+                if (!virtfn_link_file)
+                        return -ENOMEM;
+
+                if (chase_symlinks(virtfn_link_file, NULL, 0, &virtfn_pci_syspath, NULL) < 0)
+                        continue;
```
Comment (poettering) [inline src/network/networkd-sriov.c:275]: debug logging on error?

File: src/network/networkd-sriov.c [hunk 7]
```diff
@@ -103,3 +107,228 @@ int link_request_sr_iov_vfs(Link *link) {
...
+
+                if (path_equal(pci_syspath, virtfn_pci_syspath)) {
+                        found = true;
+                        break;
+                }
+        }
+        if (!found)
+                return -ENODEV;
+
+
```
Comment (poettering) [inline src/network/networkd-sriov.c:283]: spurious double newline?

File: src/network/networkd-sriov.c (hunk 5 above)
Comment (yuwata) [reply src/network/networkd-sriov.c:259]: I'd like to leave this to another later PR. At least, this sounds slightly overkill for now.

File: src/libsystemd/sd-device/sd-device.c [hunk 8]
```diff
@@ -861,6 +861,26 @@ _public_ int sd_device_get_syspath(sd_device *device, const char **ret) {
...
+
+        assert_return(ret, -EINVAL);
+        assert_return(device, -EINVAL);
+        assert_return(suffix, -EINVAL);
+
+        r = sd_device_get_syspath(device, &s);
+        if (r < 0)
+                return r;
+
+        path = path_join(s, suffix);
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c:880]: maybe refuse invalid suffixes, i.e. path_is_normalized() or so?

and maybe even make sure at least one component is specified? 

i.e. we don#t want this stuff to be used with suffixes like "../", but rather refuse that early...

File: src/network/networkd-wiphy.c [hunk 9]
```diff
@@ -214,10 +214,11 @@ int link_rfkilled(Link *link) {
         assert(link);
 
         r = link_get_wiphy(link, &w);
-        if (IN_SET(r, -EOPNOTSUPP, -ENODEV))
-                return false; /* Typically, non-wifi interface or running in container */
-        if (r < 0)
+        if (r < 0) {
+                if (r == -EOPNOTSUPP || ERRNO_IS_DEVICE_ABSENT(r))
```
Comment (poettering) [inline src/network/networkd-wiphy.c]: maybe even `ERRNO_IS_NOT_SUPPORTED() || ERRNO_IS_DEVICE_ABSETN()`?

File: src/network/networkd-wiphy.c [hunk 10]
```diff
@@ -316,14 +314,12 @@ static int wiphy_update_rfkill(Wiphy *w) {
...
+        if (!rfkill)
                 /* rfkill device may not detected by the kernel yet, and may appear later. */
-                log_wiphy_debug_errno(w, SYNTHETIC_ERRNO(ENODEV), "No rfkill device found, ignoring.");
-                return 0;
-        }
+                return -ENODEV;
 
         if (sd_device_enumerator_get_device_next(e))
-                return log_wiphy_debug_errno(w, SYNTHETIC_ERRNO(EEXIST), "Multiple rfkill devices found.");
+                return -ENXIO; /* multiple devices found */
```
Comment (poettering) [inline src/network/networkd-wiphy.c:322]: ENOTUNIQ? (just bikeshedding, ignore me)

File: src/network/networkd-wiphy.c [hunk 11]
```diff
@@ -343,12 +339,24 @@ static int wiphy_update(Wiphy *w) {
...
 
         r = wiphy_update_device(w);
-        if (r < 0)
-                return log_wiphy_debug_errno(w, r, "Failed to update wiphy device: %m");
+        if (r < 0) {
+                if (ERRNO_IS_DEVICE_ABSENT(r))
+                        log_wiphy_debug_errno(w, r, "Failed to update wiphy device, ignoring: %m");
+                else {
+                        log_wiphy_warning_errno(w, r, "Failed to update wiphy device: %m");
+                        return r;
```
Comment (poettering) [inline src/network/networkd-wiphy.c]: merge these two lines into return `log_wiphy_warning_errno(…`

File: src/network/networkd-wiphy.c [hunk 12]
```diff
@@ -343,12 +339,24 @@ static int wiphy_update(Wiphy *w) {
...
 
         r = wiphy_update_rfkill(w);
-        if (r < 0)
-                return log_wiphy_debug_errno(w, r, "Failed to update rfkill device: %m");
+        if (r < 0) {
+                if (ERRNO_IS_DEVICE_ABSENT(r))
+                        log_wiphy_debug_errno(w, r, "Failed to update rfkill device, ignoring: %m");
+                else {
+                        log_wiphy_warning_errno(w, r, "Failed to update rfkill device: %m");
+                        return r;
```
Comment (poettering) [inline src/network/networkd-wiphy.c]: ditto

**Comment by poettering** [comment]
looks good. not much clue about sriov, but codewise looks good.

**Comment by yuwata** [comment]
@poettering Thank you for the review. All comments, except for `ENOTUNIQ`, are addressed. Upgrading the green label.

**Pull request merged**